_posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md

```diff
@@ -0,0 +1,92 @@
+
+The Google Assistant directs you to use Firebase functions for request fulfillment. Firebase only supports JavaScript for a node environment -- so JavaScript it was!
+
+I never set up a local server for the Google Assistant. While the Firebase CLI is awesome and you can start a local server, debugging what is happening in that local server is not simple. I could have set up an express server equivalent to flask but I chose not to. In hindsight, I would recommend the express approach since I frustratingly ended up `console.log` debugging in the Firebase logs.
+
+### Skills as Code
+{: .text-left }
+
+I think it is a big win to have a skills training and design documented along with the implementation. Alexa knocks this out of the park with its [skill.json](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/skill.template.json) and [models](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/models/en-US.json). Developers can make changes without having to visit a website and deal with UI and UX. The ask CLI from Amazon is awesome and lets you do this and much more. You can deploy your function, update your skill and model, and test your skill. Adding a new question and training it is as simple as editing a json file. Sharing a demo or some code? Everything is documented as text.
+
+[Dialogflow](https://dialogflow.com/) is Google's machine learning natural language processing Alexa equivalent. Currently there is an API for Dialogflow but no supported clients. Everything you want to do has to go through the Dialogflow web console. The models are shared as a zip file that you have to export and import. The content within them is complicated and risky to edit by hand. I did not like having to import a zip file to look at how a demo was implemented.
+
+### Training the Model
+{: .text-left }
+
+To train Alexa to understand what to do when questions were asked was a bit cumbersome. I created a [website](https://agrc.github.io/digital-assisant-skills/) to generate Alexa utterances. Basically all of the ways a user could phrase a question. Dialogflow seemed a lot smarter and required less training. It certainly did not require a website to generate phrases. Google is winning the natural language processing in my opinion and was much easier to train.
+
+### Beyond Voice
+{: .text-left }
+
+Digital assistants lasted without screens for barely a year. Skills can now play videos, show images, and present content in basic HTML components to compliment the voice responses. Alexa has a presentation language [APL](https://developer.amazon.com/docs/alexa-presentation-language/apl-overview.html) to design for this but had limited SDK support. The Google Actions SDK for nodejs had cards, buttons, tables, images, etc all built in. It was simple and fun to add the visual components while developing. The simulator displays everything nicely.
+
+## Testing
+{: .text-left }
+
```

Comment:
```suggestion
## Testing
```

Reply:
(Sorry, didn't mean to change anything here, but didn't know how to get out of the comment box without deleting my other changes?...)